packages/viewport/src/with-viewport-match.js

```diff
@@ -32,13 +31,33 @@ import { withSelect } from '@wordpress/data';
  *
  * @return {Function} Higher-order component.
  */
-const withViewportMatch = ( queries ) => createHigherOrderComponent(
-	withSelect( ( select ) => {
-		return mapValues( queries, ( query ) => {
-			return select( 'core/viewport' ).isViewportMatch( query );
-		} );
-	} ),
-	'withViewportMatch'
-);
+const withViewportMatch = ( queries ) => {
+	const useViewPortQueriesResult = () => mapValues( queries, ( query ) => {
+		let [ operator, breakpointName ] = query.split( ' ' );
+		if ( breakpointName === undefined ) {
+			breakpointName = operator;
+			operator = '>=';
+		}
+		// Hooks should unconditionally execute in the same order,
+		// we are respecting that as from the static query of the HOC we generate
+		// a hook that calls other hooks always in the same order (because the query never changes).
+		// eslint-disable-next-line react-hooks/rules-of-hooks
+		return useViewportMatch( breakpointName, operator );
```

Comment:
Hi @youknowriad,
I did some research on this. It seems it is possible to do this. The samples people referred to were things like defining an array of states `const useInputs = [...Array(n)].map((_, i) => useState('name' + i));` and they should be avoided because normally there is a better alternative (in that case defined a single state with an array). In this specific case, I don't think we have an alternative (unless we change the hook API), and we are sure that the order the hooks are called will not change during the component life cycle, so it seems like something we could do.

Reply:
Normally, this is symptomatic of a cluttered component tree and means that you should break down the caller into multiple child components, each with their own hook call, but here we are trying to maintain backwards compatibility.

An alternative would be to introduce a `useViewportMatches` hook, but it's not a big deal. We should just clearly document that the query has to be static.

Reply:
Yes, we use this mechanism just to maintain backward compatibility. Given the way an HOC works the query will always be static for a component instance, there is no way of doing the opposite. Changing the query will require a call to the HOC and a new component is created.
I don't think we should change the hook API so I guess we can keep this solution.